### ISSUE TITLE
Tray Icon is not showing on ubuntu 18.04

### ISSUE THREAD

**Issue by leonardo2707** [z411/trackma#365]
I'm using gnome 3.28.2 and installed trackma with setup.py.

everything is working right except that.

![captura de tela de 2018-08-25 00-07-59](https://user-images.githubusercontent.com/17673896/44614504-68ffec80-a7fc-11e8-9040-5576c8355bfe.png)

**Comment by leonardo2707** [comment]
I just switched to the gnome manjaro and added the topicons extension and magically everything was perfect.

![image](https://user-images.githubusercontent.com/17673896/45334476-4c301c80-b551-11e8-8ecd-c6e6691dd82c.png)

I do not know what was wrong with the ubuntu gnome. Maybe there was a bug? or should I have forgotten to do something?

**Comment by leonardo2707** [comment]
@thebitstick nice, I will close the issue since it is working, anything I open again.

**Comment by thebitstick** [comment]
This may fix our problem but that doesn't help the problem of trackma-gtk not working. It's a bug and needs ab Issue.

**Comment by leonardo2707** [comment]
I see, I reopen and will put a bug label

**Comment by thebitstick** [comment]
Adding to this issue:
I've tested GNOME 3.30 Wayland on both Ubuntu 18.10 and Fedora 29.
trackma-gtk still does not display its tray icon under Wayland.
Running trackma-gtk under Xwayland with `GDK_BACKEND=x11` and having TopIcons Plus allows the use of tray icons.
But this should be remedied to work on Wayland.

**Comment by z411** [comment]
I'm not familiar with the direction the GNOME project is taking but isn't it because they are planning to remove tray icons altogether?

Source:
https://bugzilla.gnome.org/show_bug.cgi?id=758677

**Comment by thebitstick** [comment]
Yes, tray icons will be considered a "legacy" component for the GNOME desktop.
On Wayland, only X11 apps can use it.

So as a replacement to the tray icon, how about Trackma runs in the background, and when `trackma-gtk` is called, it'll first check for any existing processes, and ask them to unhide.

**Comment by z411** [comment]
Gnome Shell already restores the existing window if a program is re-launched so I feel that should take care of that. From what I've read about this change it seems we shouldn't be hiding the program at all. What about disabling tray icon functionality at all if we detect GNOME?

**Comment by thebitstick** [comment]
That would also work. I originally forked this repo to disable tray icons if it detects Wayland, but on second thought, I believe KDE Plasma Wayland still allows the use of "legacy" tray icons. Detecting GNOME may be difficult though. Also when detected, it would not only have to disable tray icons, but also disable "start in tray" option.

And as for

> Gnome Shell already restores the existing window if a program is re-launched so I feel that should take care of that.

Would only work if Trackma is minimized. GNOME Shell doesn't do much in that regard. Other applications that restore windows do the work themselves to detect a running instance and tell the existing one to unhide.

**Comment by thebitstick** [comment]
Also not related to tray icons but still related to GNOME:
Would a redesign featuring a HeaderBar and no menu bar be possible?

**Comment by dyskette** [comment]
> Also not related to tray icons but still related to GNOME:
> Would a redesign featuring a HeaderBar and no menu bar be possible?
> Trackma feels a bit dated on GNOME.

I would love to see trackma with headerbars. However, a well implemented headerbar needs a good design first. Because it's not only the headerbar, it's the rest of the behaviour of the application that would need to change.

More than a year ago, I played around with the design of trackma and did a bit of coding to implement it but dropped it due to lack of time.

- [trackma](https://user-images.githubusercontent.com/6687927/47969385-42500480-e03c-11e8-8b6f-9a4a75f1dba4.png)
- [trackma-preferences](https://user-images.githubusercontent.com/6687927/47969391-53007a80-e03c-11e8-9f35-5858d7fac8aa.png)

Found the mockups I did back then but they are rough...

Sorry for the offtopic. :worried:

**Comment by leonardo2707** [comment]
I had not read the official explanation before the @z411  comment.

> Official explanation:
> (https://blogs.gnome.org/aday/2017/08/31/status-icons-and-gnome/`)

I understood the point of the gnome developers, so I will close the issue since it is not a bug of the trackma.

**Comment by z411** [comment]
@thebitstick After testing it seems GtkStatusIcon won't work under Wayland KDE either so I'm just disabling tray functionality whatsoever if Wayland is detected. Please check if it works.

@dyskette By the way, did you do that mockup in an image editing program or did you code it? If the latter I might be interested in taking a look.

**Comment by thebitstick** [comment]
@z411 Fix works. GDK_BACKEND is Wayland by default and Trackma doesn't hide it self on start, and the tray icon doesn't show up. GDK_BACKEND is x11 and (TopIcons on GNOME) the tray icons show up and Trackma hides itself.

**Comment by dyskette** [comment]
@z411 I did both, but right now I just had this image mockup because the code I lost it some time in this year when I changed PC and don't know where it ended up. The code was incomplete anyway and had only only implemented a basic headerbar, but not much really.

I would be interested in helping create a version with CSD and headerbars. 😄

**Comment by lionirdeadman** [comment]
This is not a Wayland issue. Gnome *has* removed the systray but Ubuntu and others use the extension made by the Ubuntu theme to restore them. They don't use the same API though.

Extension in question : https://github.com/Ubuntu/gnome-shell-extension-appindicator

**Comment by z411** [comment]
As far as I know that extension won't work in Wayland, and when I tested it the tray icon didn't work in Wayland KDE so I assumed it was a Wayland issue, which is why I decided to look for a Wayland display as a detection measure. If you know of a reliable way to detect Gnome then I'd be happy to use that instead.

**Comment by lionirdeadman** [comment]
This extension actually does work on Gnome wayland. It simply uses another API than the old systray in Gnome hence why it didn't show up in Ubuntu 18.04 (Also would explain why Topicons which does use the old api works but they may not support Wayland, not sure)

So there's most likely 2 issues at play here :

- We don't use the new api which this extension requires hence Ubuntu 18.04 doesn't show the systray icon.

- KDE Wayland doesn't show systray icon.. Idk why, honestly. I'll go on a whim and say it's because their Wayland support is not done yet afaik.

Atleast, that's what I think. I could be wrong.
